js/scorm/wrapper.js

```diff
@@ -690,6 +677,105 @@ class ScormWrapper {
   }
 
+  getObjectiveCount() {
+    const count = this.getValue('cmi.objectives._count');
+    return count === '' ? 0 : count;
+  }
+
+  getObjectiveIndexById(id) {
+    const count = this.getObjectiveCount();
+    for (let i = 0; i < count; i++) {
+      const storedId = this.getValue(`cmi.objectives.${i}.id`);
+      if (storedId === id) return i;
+    }
+    return count;
+  }
+
+  recordObjectiveDescription(id, description) {
+    if (!this.isSCORM2004() || !description) return;
+    id = id.trim();
+    const index = this.getObjectiveIndexById(id);
+    const cmiPrefix = `cmi.objectives.${index}`;
+    this.setValue(`${cmiPrefix}.id`, id);
+    this.setValue(`${cmiPrefix}.description`, description);
+  }
+
+  recordObjectiveScore(id, score, minScore = 0, maxScore = 100, isPercentageBased = true) {
+    if (!this.isSupported('cmi.objectives._count')) {
+      this.logger.info('ScormWrapper::recordObjectiveScore: cmi.objectives are not supported by this LMS...');
+      return;
+    }
+    id = id.trim();
+    const index = this.getObjectiveIndexById(id);
+    const cmiPrefix = `cmi.objectives.${index}`;
+    this.setValue(`${cmiPrefix}.id`, id);
+    this.recordScore(cmiPrefix, score, minScore, maxScore, isPercentageBased);
+  }
+
+  recordObjectiveStatus(id, completionStatus, successStatus = 'unknown') {
+    if (!this.isSupported('cmi.objectives._count')) {
+      this.logger.info('ScormWrapper::recordObjectiveStatus: cmi.objectives are not supported by this LMS...');
+      return;
+    }
+    if (!this.isValidCompletionStatus(completionStatus)) {
+      this.handleDataError(new ScormError(CLIENT_STATUS_UNSUPPORTED, { completionStatus }));
+      return;
+    }
+    if (this.isSCORM2004() && !this.isValidSuccessStatus(successStatus)) {
+      this.handleDataError(new ScormError(CLIENT_STATUS_UNSUPPORTED, { successStatus }));
+      return;
+    }
+    id = id.trim();
+    const index = this.getObjectiveIndexById(id);
+    const cmiPrefix = `cmi.objectives.${index}`;
+    this.setValue(`${cmiPrefix}.id`, id);
+    if (this.isSCORM2004()) {
+      this.setValue(`${cmiPrefix}.completion_status`, completionStatus);
+      this.setValue(`${cmiPrefix}.success_status`, successStatus);
+      return;
+    }
+    if (completionStatus === 'completed' && successStatus !== 'unknown') completionStatus = successStatus;
+    this.setValue(`${cmiPrefix}.status`, completionStatus);
+  }
+
+  isValidCompletionStatus(status) {
+    status = status.toLowerCase(); // workaround for some LMSs (e.g. Arena) not adhering to the all-lowercase rule
+    if (this.isSCORM2004()) {
+      switch(status) {
+        case 'completed':
+        case 'incomplete':
+        case 'not attempted':
+        case 'not_attempted': // mentioned in SCORM 2004 spec - mapped to 'not attempted'
+        case 'unknown':
+          return true;
+      }
+    } else {
+      switch(status) {
+        case 'passed':
```

Comment:
These could do with being enums rather than strings. It's long winded but at least we get a properly enumerated list of the possibilities.

Reply:
Not sure if all these values are needed. Maybe `COMPLETION_STATE['NOT ATTEMPTED']` but not sure about `COMPLETION_STATE.UNKNOWN`, `COMPLETION_STATE.BROWSED`. We don't want the list to be SCORM specific, although I guess they are still generic terms. Not sure what other context they would be used for.

Reply:
You can make a new enum in SPOOR, they don't have to be added to COMPLETION_STATE in the core.